storage: clean up `pebbleIterator` construction/initialization

erikgrinaker opened the pull request:
`pebbleIterator` construction and initialization has grown a bit bloated
and convoluted over time. This patch makes some minor changes to better
express the purpose of the various options:

* `init()` now simply initializes the `pebbleIterator` by resetting the
  struct and creating the iterator options, using and reconfiguring a
  raw Pebble iterator if given.

* `newPebbleIterator()` creates a new iterator from a given Pebble
  reader.

* `newPebbleIteratorByCloning()` creates a new iterator by cloning a
  given raw Pebble iterator and reconfiguring it.

* `initReuseOrCreate()` is a convenience method specialized for batch
  iterator reuse. It (re-)initializes an existing `pebbleIterator` by
  either:

  1. Cloning and reconfiguring a given raw Pebble iterator.
  2. Using and reconfiguring a given raw Pebble iterator.
  3. Creating a new raw Pebble iterator.

Additionally, the `clonableIter` interface has been removed, in favor of
passing `pebble.Iterator` directly.

Release note: None

erikgrinaker committed:
storage: clean up `pebbleIterator` construction/initialization

`pebbleIterator` construction and initialization has grown a bit bloated
and convoluted over time. This patch makes some minor changes to better
express the purpose of the various options:

* `init()` now simply initializes the `pebbleIterator` by resetting the
  struct and creating the iterator options, using and reconfiguring a
  raw Pebble iterator if given.

* `newPebbleIterator()` creates a new iterator from a given Pebble
  reader.

* `newPebbleIteratorByCloning()` creates a new iterator by cloning a
  given raw Pebble iterator and reconfiguring it.

* `initReuseOrCreate()` is a convenience method specialized for batch
  iterator reuse. It (re-)initializes an existing `pebbleIterator` by
  either:

  1. Cloning and reconfiguring a given raw Pebble iterator.
  2. Using and reconfiguring a given raw Pebble iterator.
  3. Creating a new raw Pebble iterator.

Additionally, the `clonableIter` interface has been removed, in favor of
passing `pebble.Iterator` directly.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82789)
<!-- Reviewable:end -->

nicktrav reviewed (approved):
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82789)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @jbowens)

___
*[`pkg/storage/pebble_batch.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82789#-N4eg18tEtub6lfnfVCD:-N4eeMHg9dKWZxtctQv-:b89iy8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/2faad0a1abec2abe40ec03d08bc1a0160bde4a40/pkg/storage/pebble_batch.go#L54)):*
> ```Go
> 	iter              *pebble.Iterator
> 	writeOnly         bool
> 	iterUnused        bool
> ```

Super nitpick: some of the double negatives around `iterUnused = false` / `!iterUnused` increase the cognitive overhead. Any reason this can't be called `iterUsed`, which is `false` by default without having to set it, then flip the callsites? Then it lines up with your function sigs. Def not a blocker.


<!-- Sent from Reviewable.io -->

erikgrinaker committed:
storage: rename `iterUnused` to `iterUsed` in batch/readonly

This patch renames the field `iterUnused` to `iterUsed` in `pebbleBatch`
and `pebbleReadOnly`, flipping its meaning. This was done because the
double negative of `!p.iterUnused` to enable cloning of the iterator in
`newPebbleIterator()` could be hard to parse.

Release note: None

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82789)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens and @nicktrav)

___
*[`pkg/storage/pebble_batch.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82789#-N4eg18tEtub6lfnfVCD:-N4fe0mx9_Fa6CaxA1e9:b-ywqa44) ([raw file](https://github.com/cockroachdb/cockroach/blob/2faad0a1abec2abe40ec03d08bc1a0160bde4a40/pkg/storage/pebble_batch.go#L54)):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

Super nitpick: some of the double negatives around `iterUnused = false` / `!iterUnused` increase the cognitive overhead. Any reason this can't be called `iterUsed`, which is `false` by default without having to set it, then flip the callsites? Then it lines up with your function sigs. Def not a blocker.
</blockquote></details>

Yeah, I agree. I was a bit hesitant to start fiddling with it in case I missed any subtle interactions, but turned out to be straightforward -- added on a commit.


<!-- Sent from Reviewable.io -->
